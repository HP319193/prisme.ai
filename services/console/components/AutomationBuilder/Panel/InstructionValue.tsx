import { FC, useCallback, useMemo, useState } from 'react';
import Fieldset from '../../../layouts/Fieldset';
import { useTranslation } from 'next-i18next';
import {
  FieldProps,
  getSchemaFormLabel,
  SchemaForm,
  SchemaFormDescription,
} from '@prisme.ai/design-system';
import { CodeEditorInline } from '../../CodeEditor/lazy';
import { useField } from 'react-final-form';
import FieldContainerWithRaw from '../../FieldContainerWithRaw';
import useSchema from '../../SchemaForm/useSchema';
import { useWorkspace } from '../../../layouts/WorkspaceLayout';
import usePages from '../../PagesProvider/context';

interface InstructionValueProps {
  instruction: string;
  value: any;
  schema?: any;
  onChange: (values: any) => void;
}

const EmptyButtons: any[] = [];
const FieldAny = ({ schema, name, label }: FieldProps) => {
  const field = useField(name);
  const [value, setValue] = useState(
    typeof field.input.value === 'string'
      ? field.input.value
      : JSON.stringify(field.input.value, null, '  ')
  );
  const onChange = useCallback(
    (value: string) => {
      setValue(value);
      try {
        const json = JSON.parse(value);
        field.input.onChange(json);
      } catch {
        field.input.onChange(value);
      }
    },
    [field.input]
  );

  return (
    <div className="flex flex-1 flex-col my-2">
      <SchemaFormDescription text={schema.description}>
        <label className="text-[10px] text-gray">
          {label || schema.title || getSchemaFormLabel(name)}
        </label>
        <CodeEditorInline
          value={value}
          onChange={onChange}
          mode="json"
          style={{ flex: 'auto' }}
        />
      </SchemaFormDescription>
    </div>
  );
};

const components = {
  FieldAny,
  FieldContainer: FieldContainerWithRaw,
};

export const InstructionValue: FC<InstructionValueProps> = ({
  instruction,
  value,
  schema = {},
  onChange,
}) => {
  const { workspace } = useWorkspace();
  const { pages } = usePages();
  const { t } = useTranslation('workspaces');

  const appInstance = useMemo(() => {
    if (!workspace.imports) return workspace.config;
    const [appName] = instruction.split(/\./);
    if (!workspace.imports[appName]) return workspace.config;
    return workspace.imports[appName].config || {};
  }, [instruction, workspace.config, workspace.imports]);

  const { extractSelectOptions } = useSchema({
    config: appInstance,
    automations: workspace.automations,
    pages: pages.get(workspace.id),
  });

  const cleanedSchema = useMemo(() => {
    const cleaned = {
      ...schema,
      title: t('automations.instruction.description', {
        context: instruction,
        default: schema.description,
      }),
    };
    if (instruction === 'repeat') {
      cleaned['ui:options'] = {
        oneOf: {
          options: [
            {
              label: 'automations.instruction.form.repeat.on.label',
              index: 0,
              value: {
                until: undefined,
              },
            },
            {
              label: 'automations.instruction.form.repeat.until.label',
              index: 1,
              value: {
                on: undefined,
              },
            },
          ],
        },
      };
    }
    return cleaned;
  }, [instruction, schema, t]);

  const locales = useMemo(
    () => ({
      addItem: t('automations.instruction.form.addItem'),
      addProperty: t('automations.instruction.form.addProperty'),
      propertyKey: t('automations.instruction.form.propertyKey'),
      propertyValue: t('automations.instruction.form.propertyValue'),
      removeItem: t('automations.instruction.form.removeItem'),
      removeProperty: t('automations.instruction.form.removeProperty'),
      uploadLabel: t('automations.instruction.form.uploadLabel'),
      uploadRemove: t('automations.instruction.form.uploadRemove'),
    }),
    [t]
  );

  if (!schema) return null;

  return (
    <Fieldset
      legend={t('automations.instruction.label', { context: instruction })}
      hasDivider={false}
    >
      <SchemaForm
        schema={cleanedSchema}
        onChange={onChange}
        initialValues={value}
        buttons={EmptyButtons}
        components={components}
        locales={locales}
        utils={{ extractSelectOptions }}
      />
    </Fieldset>
  );
};

export default InstructionValue;
