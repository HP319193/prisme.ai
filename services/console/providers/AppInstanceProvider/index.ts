export * from './AppInstanceProvider';
export { default } from './AppInstanceProvider';
