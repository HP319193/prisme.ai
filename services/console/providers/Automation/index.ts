export * from './AutomationProvider';
export { default } from './AutomationProvider';
