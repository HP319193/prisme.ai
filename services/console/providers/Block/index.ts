export * from './BlockProvider';
export { default } from './BlockProvider';
