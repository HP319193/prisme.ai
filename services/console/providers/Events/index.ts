export * from './EventsProvider';
export { default } from './EventsProvider';
