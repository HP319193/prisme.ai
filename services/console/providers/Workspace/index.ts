export * from './WorkspaceProvider';
export { default } from './WorkspaceProvider';
