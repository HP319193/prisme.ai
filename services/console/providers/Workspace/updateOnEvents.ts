import { Dispatch, SetStateAction } from 'react';
import { Events } from '../../utils/api';
import { Workspace } from './WorkspaceProvider';

type EventCallback = (workspace: Workspace, payload: any) => Workspace;
type EventsListeners = Record<string, EventCallback>;

function insertItemInWorkspaceList(
  workspace: Workspace,
  list: string,
  slug: string,
  item: any,
  prevSlug: string = slug
) {
  const currentList = workspace[list as keyof Workspace] || {};
  const exists = !!currentList[prevSlug as keyof typeof currentList];

  return {
    ...workspace,
    [list]: {
      ...Object.entries(workspace[list as keyof Workspace] || {}).reduce(
        (prev, [s, p]) => ({
          ...prev,
          [s === prevSlug ? slug : s]:
            s === prevSlug
              ? {
                  ...p,
                  ...item,
                }
              : p,
        }),
        {}
      ),
      ...(exists ? {} : { [slug]: item }),
    },
  };
}

function removeItemInWorkspaceList(
  workspace: Workspace,
  list: string,
  slug: string
) {
  return {
    ...workspace,
    [list]: Object.entries(workspace[list as keyof Workspace] || {}).reduce(
      (prev, [s, p]) =>
        s === slug
          ? prev
          : {
              ...prev,
              [s]: p,
            },
      {}
    ),
  };
}

const eventsListeners: EventsListeners = {
  'workspaces.automations.updated': (
    workspace,
    { slug, oldSlug, automation }
  ) =>
    insertItemInWorkspaceList(
      workspace,
      'automations',
      slug,
      automation,
      oldSlug
    ),
  'workspaces.automations.deleted': (workspace, { automationSlug }) =>
    removeItemInWorkspaceList(workspace, 'automations', automationSlug),
  'workspaces.pages.updated': (workspace, { slug, oldSlug, page }) =>
    insertItemInWorkspaceList(workspace, 'pages', slug, page, oldSlug),
  'workspaces.pages.deleted': (workspace, { pageSlug }) =>
    removeItemInWorkspaceList(workspace, 'pages', pageSlug),
  'workspaces.apps.configured': (workspace, { slug, oldSlug, appInstance }) =>
    insertItemInWorkspaceList(workspace, 'imports', slug, appInstance, oldSlug),
  'workspaces.apps.uninstalled': (workspace, { slug }) =>
    removeItemInWorkspaceList(workspace, 'imports', slug),
};

export const updateOnEvents = (
  events: Events,
  setWorkspace: Dispatch<SetStateAction<Workspace | undefined>>
) => {
  Object.entries(eventsListeners).forEach(([event, callback]) => {
    events.on(event, ({ payload }) => {
      setWorkspace((workspace) => workspace && callback(workspace, payload));
    });
  });
};

export default updateOnEvents;
