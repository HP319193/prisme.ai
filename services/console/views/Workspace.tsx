import {
  Button,
  Col,
  MenuTab,
  Row,
  SidePanel,
  Tooltip,
} from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import { useEffect, useMemo, useState } from 'react';
import { EventsViewer } from '../components/EventsViewer';
import getLayout from '../layouts/WorkspaceLayout';
import AutomationsSidebar from './AutomationsSidebar';
import IconApps from '../icons/icon-apps.svgr';
import IconAutomations from '../icons/icon-automations.svgr';
import IconPages from '../icons/icon-pages.svgr';
import AppsSidebar from './AppsSidebar';
import PagesSidebar from './PagesSidebar';
import { DoubleLeftOutlined, DoubleRightOutlined } from '@ant-design/icons';
import ShareWorkspace from '../components/Share/ShareWorkspace';
import Storage from '../utils/Storage';
import { useWorkspace } from '../components/WorkspaceProvider';

export const Workspace = () => {
  const { t } = useTranslation('workspaces');
  const { workspace, fullSidebar, setFullSidebar, setShare } = useWorkspace();
  const [sidebar, setSidebar] = useState(
    Storage.get('__workpaceSidebar') || 'automations'
  );

  useEffect(() => {
    Storage.set('__workpaceSidebar', sidebar);
  }, [sidebar]);

  useEffect(() => {
    setShare({
      label: t('workspace.share'),
      component: ShareWorkspace,
    });
  }, [setShare, t]);

  const menu = useMemo(
    () => [
      {
        label: (
          <div className="flex items-center">
            <div className="flex mr-2">
              <IconApps width={16} height={16} />
            </div>
            {t('apps.link')}
          </div>
        ),
        key: 'apps',
      },
      {
        label: (
          <div className="flex items-center">
            <div className="flex mr-2">
              <IconAutomations width={16} height={16} />
            </div>
            {t('automations.link')}
          </div>
        ),
        key: 'automations',
      },
      {
        label: (
          <div className="flex items-center">
            <div className="flex mr-2">
              <IconPages width={16} height={16} />
            </div>
            {t('pages.link')}
          </div>
        ),
        key: 'pages',
      },
    ],
    [t]
  );
  console.log('wp', workspace);

  if (!workspace) return null;

  return (
    <>
      <Row className="h-full">
        <Col span={14} className="flex h-full">
          <EventsViewer />
        </Col>
        <Col span={10} className="flex h-full">
          <SidePanel
            variant="squared"
            className={`
            absolute top-0 right-0 bottom-0 !bg-white
            ${fullSidebar ? 'w-[90vw] drop-shadow' : 'w-full'}
            transition-all ease-in duration-200`}
            Header={
              <div className="flex flex-row items-center h-[70px] justify-between border border-gray-200 border-solid !border-t-0">
                <MenuTab
                  items={menu}
                  selected={
                    (menu.find(({ key }) => sidebar === key) || menu[0]).key
                  }
                  onSelect={setSidebar}
                />
                <Tooltip
                  title={t(
                    fullSidebar
                      ? 'workspace.collapseSidebar'
                      : 'workspace.expandSidebar'
                  )}
                  placement="left"
                >
                  <Button
                    onClick={() => setFullSidebar(!fullSidebar)}
                    className="!text-sm"
                  >
                    {fullSidebar ? (
                      <DoubleRightOutlined
                        className="color-blue"
                        alt="workspace.collapseSidebar"
                      />
                    ) : (
                      <DoubleLeftOutlined
                        className="color-blue"
                        alt="workspace.expandSidebar"
                      />
                    )}
                  </Button>
                </Tooltip>
              </div>
            }
          >
            <>
              {sidebar === 'apps' && <AppsSidebar />}
              {sidebar === 'automations' && <AutomationsSidebar />}
              {sidebar === 'pages' && <PagesSidebar />}
            </>
          </SidePanel>
        </Col>
      </Row>
    </>
  );
};

Workspace.getLayout = getLayout;

export default Workspace;
