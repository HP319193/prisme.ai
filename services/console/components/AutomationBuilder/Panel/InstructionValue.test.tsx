import InstructionValue from './InstructionValue';
import renderer from 'react-test-renderer';
import { workspaceContext } from '../../../providers/Workspace';

jest.mock('../../../utils/useYaml', () => {
  const toJSON = jest.fn();
  const toYaml = jest.fn();
  const useYaml = jest.fn(() => ({
    toJSON,
    toYaml,
  }));
  return useYaml;
});

it('should render', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <workspaceContext.Provider value={{ workspace: { imports: {} } } as any}>
      <InstructionValue
        instruction="emit"
        schema={{ properties: { event: { type: 'string' } } }}
        value={{}}
        onChange={onChange}
      />
    </workspaceContext.Provider>
  );
  expect(root.toJSON()).toMatchSnapshot();
});

it('should not render without schema', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <workspaceContext.Provider value={{ workspace: { imports: {} } } as any}>
      <InstructionValue instruction="emit" value={{}} onChange={onChange} />
    </workspaceContext.Provider>
  );
  expect(root.toJSON()).toMatchSnapshot();
});
