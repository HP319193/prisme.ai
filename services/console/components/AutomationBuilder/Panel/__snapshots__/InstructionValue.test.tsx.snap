// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render without schema 1`] = `null`;

exports[`should render 1`] = `
<div
  className="mt-5"
>
  emit
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-5"
    >
      <span
        className="flex flex-col "
      />
    </div>
    automations.edit.save
  </form>
</div>
`;
